### ISSUE TITLE
Diagnostic enhancement: "auto" in lambda expression under -std=c++11

### ISSUE THREAD

**Issue by haoxintu** [llvm/llvm-project#46059]
|  |  |
| --- | --- |
| Bugzilla Link | [46714](https://llvm.org/bz46714) |
| Version | trunk |
| OS | Linux |
| CC | @dwblaikie,@zygoloid |

## Extended Description 
Hi, all.

"auto" used in lambda expression is a c++14 extension, but clang doesn't emit the appropriate diagnostic message about it.

For example,

```cpp
$cat test.cc
int main(){
    auto var = [](auto){};
    return 0;
}
```

```
$clang++ -std=c++11 -c test.cc
test.cc:2:19: error: 'auto' not allowed in lambda parameter
    auto var = [](auto){};
                  ^~~~
1 error generated.
```

I guess clang should give a more specific diagnostic about this like GCC.

```
$g++ -std=c++11 -c test.cc
test.cc: In function ‘int main()’:
test.cc:2:19: error: use of ‘auto’ in lambda parameter declaration only available with ‘-std=c++14’ or ‘-std=gnu++14’
    2 |     auto var = [](auto){};
      |                   ^~~~
```

Just a small enhancement I guess, please understand if anything I stated is unsuitable.

Thanks,
Haoxin

**Comment by Endilll** [comment]
@cor3ntin can you point out where this diagnostic is defined in sources? I haven't been able to find it.

**Comment by cor3ntin** [comment]
The code which determines there is an error is here
https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/SemaType.cpp#L3584-L3594

The error is produced a bit below
https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/SemaType.cpp#L3737

A solution might be to keep track of when the errors are conditioned on language versions, and modify the existing diagnostic to add "before C++20"

**Comment by Endilll** [comment]
https://github.com/llvm/llvm-project/blob/fa2b038cadf17d08014e5fb75c47b5024860953e/clang/include/clang/Basic/DiagnosticSemaKinds.td#L2345-L2360

I've tried to get through this even before writing my previous comment, but I wasn't able to recognize the message, and find it hard still.

CC @AaronBallman

**Comment by AmrDeveloper** [comment]
@cor3ntin, Issue should be closed as completed in this [commit](https://github.com/llvm/llvm-project/commit/476f7f65f9f17fab7e78f395b83dcb7b0bbd5215)
